src/appleseed/foundation/math/correlatedmultijitter.h

```diff
@@ -0,0 +1,158 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+#ifndef APPLESEED_FOUNDATION_MATH_CMJ_H
+#define APPLESEED_FOUNDATION_MATH_CMJ_H
+
+// appleseed.foundation headers.
+#include "foundation/math/vector.h"
+
+// Standard headers.
+#include <cassert>
+#include <cmath>
+
+namespace foundation
+{
+
+//
+// Correlated Multi-Jittered Sampling.
+//
+// References:
+//
+//   https://graphics.pixar.com/library/MultiJitteredSampling/paper.pdf
+//
+
+inline void cmj_compute_mn(unsigned int & m, unsigned int & n, const unsigned int N);
+
+template <typename T>
+inline Vector<T, 2> cmj_generate_sample(
+    unsigned int        s,
+    const unsigned int  m,
+    const unsigned int  n,
+    const unsigned int  N,
+    const unsigned int  p);
+
+template <typename T>
+inline Vector<T, 2> cmj_generate_sample(
+    unsigned int        s,
+    const unsigned int  N,
+    const unsigned int  p);
+
+inline unsigned int cmj_permute(
+    unsigned int i,
+    const unsigned int l,
+    const unsigned int p);
+
+inline float cmj_randfloat (unsigned int i, const unsigned int p);
+
+
+//
+// Correlated Multi-Jittered Sampling implementation.
+//
+
+inline void cmj_compute_mn(unsigned int & m, unsigned int & n, const unsigned int N)
+{
+    m = static_cast<int>(std::sqrt((float) N));
+    n = (N + m - 1) / m;
+}
+
+template <typename T>
+inline Vector<T, 2> cmj_generate_sample(
+    unsigned int        s,
+    const unsigned int  m,
+    const unsigned int  n,
+    const unsigned int  N,
+    const unsigned int  p)
+{
+    assert(m > 0 && n > 0);
+    s = cmj_permute(s, N, p * 0x51633e2d);
+    int sx = cmj_permute(s % m, m, p * 0x68bc21eb);
+    int sy = cmj_permute(s / m, n, p * 0x02e5be93);
+    float jx = cmj_randfloat(s, p * 0x967a889b);
+    float jy = cmj_randfloat(s, p * 0x368cc8b7);
+    Vector<T, 2> r;
+    r[0] = (sx + (sy + jx) / n) / m;
+    r[1] = (s + jy) / N;
+    return r;
+}
+
+template <typename T>
+inline Vector<T, 2> cmj_generate_sample(
```

Comment:
Unused function

Reply:
No but can be usefull. Should I remove it ?

Reply:
1) this is really easy to rewrite this one if it is needed
2) in my experience a lot of bugs comes from using dead code